library/core/src/iter/adapters/intersperse.rs

```diff
@@ -11,16 +12,25 @@ where
     I::Item: Clone,
 {
     separator: I::Item,
-    iter: Peekable<I>,
-    needs_sep: bool,
+    next_item: Option<I::Item>,
+    iter: Fuse<I>,
+}
+
+#[unstable(feature = "iter_intersperse", reason = "recently added", issue = "79524")]
+impl<I> FusedIterator for Intersperse<I>
+where
+    I: FusedIterator,
+    I::Item: Clone,
+{
 }
 
 impl<I: Iterator> Intersperse<I>
 where
     I::Item: Clone,
 {
     pub(in crate::iter) fn new(iter: I, separator: I::Item) -> Self {
-        Self { iter: iter.peekable(), separator, needs_sep: false }
+        let mut iter = iter.fuse();
+        Self { separator, next_item: iter.next(), iter }
```

Comment:
@m-ou-se thx, do you know if this is a documented expectation, or just a convention that most iterators happen to adhere to?  I have a suspicion that handling this edge case may have a performance penalty for some iterators - essentially any "peekable"-like iterator would need to store `Option<Option<I::Item>>` instead of `Option<I::Item>` to know if the iteration has started, and do an extra condition on each iteration. I will add it to my code, but the 2x performance gain might be significantly less if any.